### ISSUE TITLE
Fix `TryConvWith` for SourceEdit

### ISSUE THREAD

**Issue by matklad** [rust-analyzer/rust-analyzer#105]
TBD

**Comment by vemoo** [comment]
I've started to take a look at this here https://github.com/vemoo/rust-analyzer/tree/translate_offset_with_edit_proptest

I have some questions about how `pre_edit_index`, `offset` and `edits` are related.

- Are `offset` and `edits` required to reference positions and ranges inside `pre_edit_index` or should `translate_offset_with_edit` handle invalid offset and edits?
- At the moment `translate_offset_with_edit` only looks at the first edit. Should it check all of them?

**Comment by matklad** [comment]
Just a quick comment now: Yep, it should handle all edits. Check the lsp docs for semantics of several edits

**Comment by matklad** [comment]
1) the *first* edit uses the same offsets as in the pre_edit_index. The second edit uses the coordinate system where the first edit is already applied, the third one — where the first two edits are applied. Offset uses the coordinate system after all the edits are applied.

2) yep, it should check for all edits.

**Comment by vemoo** [comment]
Acording to this: https://microsoft.github.io/language-server-protocol/specification#textedit the edit ranges cannot overlap and all refer to positions in the original text.
Are `AtomEdit`s the same kind of edits? Or are they an internal representation?

**Comment by matklad** [comment]
Oh, well, I am confusing `TextEdit[]` with `TextDocumentContentChangeEvent[]`

https://microsoft.github.io/language-server-protocol/specification#textDocument_didChange

(yep, LSP has *both* semantics in the same protocol :unamused: )

Without looking at the code, I actually can't say which semantic is used by our `AtomicEdit`, but I *guess* that it is "edit ranges cannot overlap and all refer to positions in the original text.". 

I think that means that we could use some internal refactorings first, to make things clear and to have a good place to store proptest impls. 

Let's introduce `ra_text_edit` crate, which stores `AtomicEdit` (wich is currently in ra_syntax) and `Edit` (which is in `ra_editor`). It should depend only on `text_unit` I guess? It'll probably be a good idea to rename them to `AtomicTextEdit` and `TextEdit` in process (to distinguish from edits across several files). In general, most of the code should use `TextEdit`, which should document semantics of atomic edits and enforce invariants during construnction.

**Comment by matklad** [comment]
>Are AtomEdits the same kind of edits? Or are they an internal representation?

The major theme of rust analyzer is to keep it mostly independent from LSP. So, ra_analys, the main API, is just a rust library with convenient to Rust data strctures. ra_lsp_server converts types from ra_analysis  to LSP types. That is quite a bit more boilerplate then using LSP types directly, but I think this is a good approach. 

So, `AtomicEdit` is a public type of `ra_analysis`: it doesn't have to match any specific LSP type, it just needs to be convenient to use from Rust.

**Comment by vemoo** [comment]
> Let's introduce `ra_text_edit` crate, which stores `AtomicEdit` (wich is currently in ra_syntax) and `Edit` (which is in `ra_editor`). It should depend only on `text_unit` I guess? It'll probably be a good idea to rename them to `AtomicTextEdit` and `TextEdit` in process (to distinguish from edits across several files). In general, most of the code should use `TextEdit`, which should document semantics of atomic edits and enforce invariants during construnction.

Ok, I will start working on this.

**Comment by vemoo** [comment]
I've been thinking about the semantics of `AtomEdit`s, and if they represent unsaved changes it makes sense that they must not overlap and all refer to positions in the original text.

Also creating arbitrary edits it's a little trickier than i thought, because which edits are valid depend on the previously created edits.

**Comment by matklad** [comment]
>Also creating arbitrary edits it's a little trickier than i thought, because which edits are valid depend on the previously created edits.

Yep, that is indeed tricky. Take a look at similar code in fall: https://github.com/matklad/fall/blob/639506523ef72c207f4a79d1ac12bb952771334d/fall/text/src/prop.rs#L17-L34

The way it works is that it creates a sorted list of disjoint segments, "cuts", by creating a sorted list of points and grouping them in pairs. Note, however, that that code is also wrong, because it does not generate valid text offsets, because it doesn't have an input string, while it should have one.

**Comment by vemoo** [comment]
I ended up doing something similar: https://github.com/vemoo/rust-analyzer/blob/36ed81e92623831579b2654c085edf536b603a55/crates/ra_text_edit/src/test_utils.rs#L26-L73
But i use [`proptest::sample::subsequence`](https://altsysrq.github.io/rustdoc/proptest/0.8.7/proptest/sample/fn.subsequence.html) because it returns the elements in same order as the collection and non contiguous. As is, it's too slow, so i'll try to simplify it, but i think it's pretty complete.

I think using `prop::collection::vec` could give invalid edits because it could return for example: 3, 3, 3, 3 which end up being two empty ranges and i think that would be considered overlapping?

**Comment by matklad** [comment]
Looks reasonable to me!

**Comment by vemoo** [comment]
It took me longer than I expected but I finally managed. There's a few things that I had to do.

- To be able to use proptest impls across crates I had to add `proptest` as a normal dependecy and make them public
- After working on it for a bit it made more sense to put it in `ra_editor` because it's closely related to `LineIndex`
- The aproach I took was to iterate over all elements which can affect the line-col mapping. I iterate over the original newlines and multibyte chars checking each time if there's an edit that deletes or moves each element in the original, I also iterate over the inserted newlines and multibyte chars. When I find something past the offset I'm trying to translate I return.
This requires that the edits are in order and for now I allocate a `Vec` and insert the in order there, but ther's no other allocation, and this allocation could be avoided if they are sorted.
- I added benchmarks to verify that it was faster than editing the text and counting lines and columns. It's faster but not by much
- I found a small mistake in `LineIndex` where the multibyte chars after the last newline where being ignored

I'm going to add proptest based test to `LineIndex` but the rest is already here: https://github.com/vemoo/rust-analyzer/tree/ra-text-edit-refactor. Let me know if you see something that should be changed.
Should I clean-up the commit history before opening a pull-request?
